### ISSUE TITLE
Enhancement: [@typescript-eslint/no-magic-numbers] Allow bit mask initializer for enum

### ISSUE THREAD

**Issue by yf-yang** [typescript-eslint/typescript-eslint#8001]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-magic-numbers/

### Description

Although no-magic-numbers has an option `ignoreEnums`, it won't ignore bit masks in the enum initializer. I am not sure what is a good way to deal with it, I'd prefer to directly allow bit masks when `ignoreEnum` is enabled. Another solution would be add another option to allow it.

[playground link](https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=KYOwrgtgBAolDeAoKUCCUC8UCMUA8eUADADTJQBCmO%2Bh2ZKAwtbgVAEwNQAiLtUAZjIBfIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFsBDAc0plaTWNwBGiaKgwBtcNhxToHaJAA0CxVkXZIlfp2iIAoqO4yw%2BOIk269Bo4nRhZABnVgAjAF0t2AF9-PwgggKA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

### Fail

```typescript
no example
```

### Pass

```typescript
enum E {
  A = 1 << 0,
  B = 1 << 1,
  C = 1 << 2,
  D = 1 << 3,
}
```

### Additional Info

The implementation could be similar to https://github.com/typescript-eslint/typescript-eslint/pull/3515

**Comment by bradzacher** [comment]
My stance here is that we shouldn't deviate our logic from what the base rule uses to flag magic numbers.

For all cases the base rule flags bitshift operations - so I think that's the logic we should use.
As this is a stylistic rule - there's no chance that they will change the upstream logic to add an option for bitshifts either as stylistic rules are frozen upstream.

I did some searching and I can't even find an issue in ESLint core asking for the addition of an option to ignore bitwise operators or similar. Which means that likely the intersection of codebases that use no-magic-numbers and use bitshift operators is tiny.
TBH usage of the rule in general is minimal because it's so cumbersome to use.

The rule was added to our project in may 2019 (I.e. Over 4 years ago) and this is the first request for such a feature - which further shows that the impacted users is very low.

For these reasons I'm going to close this as a "won't fix"

**Comment by yf-yang** [comment]
Array.includes makes sense 🤔, I'll take a try.
Still, I'm curios, why does https://github.com/typescript-eslint/typescript-eslint/issues/3487 get merged into the codebase?

**Comment by bradzacher** [comment]
That's a very different problem to this one!

That rule is our own rule, not an extension rule - so we're not looking to keep in sync with upstream.

The rule also has a different concern to this rule - this rule is about avoiding aliasing of numbers, whereas that rule is about enforcing preventing the usage of variables in enums to avoid errors related to shadowing.
